### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by danielgtaylor**
This package provides a serializer for API Blueprint as a Fury.js adapter. It copies the serializer component from Fury.js along with the test fixtures available there. It also switches from the now-defunct [Swig](https://github.com/paularmstrong/swig/issues/628) to [Nunjucks](https://mozilla.github.io/nunjucks/) for the template engine.

In the future we may want to consider a repo of test fixtures for serializers to use, but for now I think this will be okay.

The tests are going to fail because they depend on the `fury` package to load elements from fixtures and will need apiaryio/fury.js#48. Not sure what I can do about that :tired_face: :chicken: :egg:

### CODE REVIEW

File: src/template.nunjucks [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+{% else %}
+  {% if example.headers and example.headers.exclude('Content-Type').length %}
+    + Headers
+
+      {% for header in example.headers.exclude('Content-Type').content %}
+            {{ header.key.toValue() }}: {{ header.value.toValue() }}{% endfor %}
+
+  {% endif %}
+  {% if example.dataStructure %}
+    {{ example.dataStructure.content|mson|indent(4) }}
```
Comment (kylef) [inline src/template.nunjucks:24]: Is this missing a `+ Attributes` before the attributes?

File: src/template.nunjucks [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+
+
+
+{% macro renderParameters(hrefVariables) %}
++ Parameters
+
+  {% for item in hrefVariables.content %}
+    {% set typeAttributes = item.attributes.get('typeAttributes') %}
+    + {{ item.key.toValue() }}{% if typeAttributes %} ({{ typeAttributes.toValue() }}){% endif %}{% if item.value.description %} - {{ item.value.description }}{% endif %}{% endfor %}
```
Comment (kylef) [inline src/template.nunjucks]: Unless I'm missing something, this doesn't appear to show the item's value? I.e:

```
+ Parameters
    + id: 5 (number)
```

Comment (kylef) [inline src/template.nunjucks]: Unless I'm missing something, this doesn't appear to show the item's value? I.e:

```apib
+ Parameters
    + id: 5 (number)
```

Comment (kylef) [inline src/template.nunjucks]: Unless I'm missing something, this doesn't appear to show the item's value? I.e:

``` apib
+ Parameters
    + id: 5 (number)
```

File: src/adapter.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+env.addFilter('mson', mson);
+env.addFilter('indent', indent);
+env.addFilter('bodyOnly', bodyOnly);
+env.addFilter('resourceShorthand', resourceShorthand);
+env.addFilter('pretty', pretty);
+env.addFilter('getCopy', getCopy);
+
+export const name = 'api-blueprint-serializer';
+export const mediaTypes = [
+  'text/vnd.apiblueprint',
```
Comment (kylef) [inline src/adapter.js:23]: Shouldn't this be `text/vnd.apiblueprint+markdown`?

ref https://github.com/apiaryio/apiblueprintorg/blob/master/apiary.apib#L15

File: src/template.nunjucks (hunk 1 above)
Comment (danielgtaylor) [inline src/template.nunjucks:24]: The `mson` filter automatically adds it.

File: src/adapter.js (hunk 3 above)
Comment (danielgtaylor) [inline src/adapter.js:23]: Good question. Honestly I'm not sure, since it isn't defined in the API Blueprint spec. Is there an official registered mimetype somewhere?

Comment (kylef) [inline src/adapter.js:23]: It's not registered, `text/vnd.apiblueprint+markdown` is what's used in api.apiblueprint.org and Hyperdrive.

@zdne can you perhaps confirm the correct content-type for API Blueprint in markdown form?

File: src/template.nunjucks (hunk 1 above)
Comment (kylef) [inline src/template.nunjucks:24]: Thanks for clarifying. :white_check_mark:

File: src/adapter.js (hunk 3 above)
Comment (danielgtaylor) [inline src/adapter.js:23]: And to make matters even more confusing, [Deckard Cain](https://github.com/apiaryio/deckardcain) will return [`'text/vnd.apiblueprint'`](https://github.com/apiaryio/deckardcain/blob/master/src/deckardcain.es6#L19) when detecting input formats...

File: src/template.nunjucks (hunk 2 above)
Comment (danielgtaylor) [inline src/template.nunjucks]: Latest push adds support for parameter sample values. Hopefully in the future this code path can be merged with the full MSON code.

File: src/adapter.js (hunk 3 above)
Comment (zdne) [inline src/adapter.js:23]: `text/vnd.apiblueprint` it is (I would like to move away from `+markdown` where possible.

File: package.json [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    "babel-runtime": "^5.8.20",
+    "nunjucks": "^2.0.0",
+    "peasant": "^0.5.2"
+  },
+  "devDependencies": {
+    "chai": "^3.2.0",
+    "coveralls": "^2.11.2",
+    "fury": "^0.8.4",
+    "glob": "^5.0.14",
+    "peasant": "^0.5.0"
```
Comment (pksunkara) [inline package.json]: Peasant to be updated to `0.5.2`?

File: package.json [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    "ci": "peasant -s lint test build",
+    "prepublish": "npm run ci",
+    "cover": "peasant cover",
+    "peasant": "peasant",
+    "lint": "peasant lint"
+  },
+  "dependencies": {
+    "babel-runtime": "^5.8.20",
+    "nunjucks": "^2.0.0",
+    "peasant": "^0.5.2"
```
Comment (pksunkara) [inline package.json]: Why is this in `dependencies`?

File: .travis.yml [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+- '0.12'
+- '0.10'
+script:
+- npm run ci
+after_script:
+- npm run coveralls
+notifications:
+  email: false
+  slack:
+    secure: bUlFDXuxqBjOKweX0sO6RRKKHnwFFoqKxffv+O5Ivti8lFOKnFonLvyH6PBJms3+nqHi7gLGx6cs4mAesdaFIC5pROA7Qcod/KS8joqvpxZGIloZRhuVin8J8ly9C89lVIvgPpL9MqrvQf/212EnfEcjTTzQdbK62A35Ysygat0hQefIz99+FhO6sRW5fiyR8q5QcxQ1uwauHgP6W4xnbqxJ8sIIPi9DQhCErU6xUr2sPYPaIs0dnMKCQIJpCAFZou2MIr7aTzpw9Hn/orgamHR2yEntWp5nEZykJw5bW/UKj5Z9JkpgxL7NEj+sPhQkq0jpzrvJSRssxPVvXCJAE++ocConLtV0h1BV7qX2ftKYg5Gw6ZoqybYzReX9u7mzw68cRdWAzKqhhzYC/8rv/nSyyu2yC+AMZ+jnDSunlto6B4/8GDV06uu1c2UMC667fs9iZXiviy963X7mx1ZvMinNbJMqFuJ2Hwd3Hhnvx+Gd97DuXqdmNZCGYeAtB/cc+mZqsrorc+I9A57cbZIR+C2Qk8TLvM3lysznZ8rub4DwMu3UvEGXIEln83R1vjLQmYabDyFZUK5TSQrYgXpwo3Fe0tfnEb6z7pJIGXvEE45vcJKeKVQaGE2n3CW5BM99wtU8/HjIzqFIQ8YIP/34hZ+Hv/9LTJTKapGUZEK0KSQ=
```
Comment (pksunkara) [inline .travis.yml]: Please update this to be specific to this repository

File: src/mson.js [hunk 7]
```diff
@@ -0,0 +1,238 @@
+/*
+ * Renders refract elements into MSON.
+ */
+
+/*
+ * Indent a piece of multiline text by a number of spaces.
+ * Setting `first` to `true` will also indent the first line.
+ */
+function indent(input, spaces, options = {first: false}) {
```
Comment (pksunkara) [inline src/mson.js]: Isn't https://github.com/apiaryio/fury-adapter-apib-serializer/pull/1/files#diff-e52a9a67956bd51d044af4058fb939ceR16 just a repeat of this?

Comment (danielgtaylor) [inline src/mson.js]: Good catch. These were written at different times and do the exact same thing, so I've refactored the code to reuse a single function.

File: src/mson.js [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+    } else if (item.element === 'ref') {
+      renderedContent += `${marker} Include ${item.content.href}\n`;
+    } else if (item.element === 'select') {
+      // This is a `OneOf` mutually exclusive type. Currently not
+      // supported as it needs some support upstream in Minim.
+      console.warn('MSON select/option elements are not yet supported!');
+    } else {
+      // This is an array type or something similar.
+      objectLike = false;
+      /* eslint-disable no-use-before-define */
```
Comment (pksunkara) [inline src/mson.js:66]: Do we have to do this a lot? I know this is something about disabling a feature in linting. But I would love to know more about it.

File: src/template.nunjucks [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+{% endmacro %}
+
+
+
+
+
+{% macro renderResource(resource) %}
+{% if not resource|resourceShorthand %}
+{% set href = null %}
+### {{ resource.title|default('Resource', true) }} [{{ resource.href }}]
```
Comment (pksunkara) [inline src/template.nunjucks]: We shouldn't use the default `Resource` here. The following is a valid blueprint.

``` apib
# Tesla

# /users

# List [GET]
+ Response 204
```

**Comment by pksunkara** [comment]
Just one issue that needs to be fixed.

Regarding the tests, I am afraid we have to do something about it. Maybe rip out the legacy interface into another module? Let's discuss on slack tomorrow.

File: src/mson.js (hunk 8 above)
Comment (kylef) [inline src/mson.js:66]: http://eslint.org/docs/rules/no-use-before-define.html

Comment (danielgtaylor) [inline src/mson.js:66]: @pksunkara, the specific problem here is that we essentially have recursion through a call loop:

``` js
function a(value) {
  if (value >= 100) { return value; }
  return b(value);                    // <--- at this point, `b` is not yet defined
}

function b(value) {
  return a(value + 1);
}

a(5);
```

In the function `a` it calls an undefined function `b`, but we can't define it at the top via `var b;` because you can't do that **plus** create a new function with the same name (redefining a variable is also a lint error). Because of that, we just shut off the rule for those couple lines. It's really kind of a special edge case.

File: src/template.nunjucks (hunk 9 above)
Comment (danielgtaylor) [inline src/template.nunjucks]: Nice catch. Updated.

**Comment by danielgtaylor** [comment]
@pksunkara the tests can work with the latest Fury patch release, but only on Node.js 0.10 because of the Robotskirt dependency. I've updated the Travis config to reflect that and we'll uncomment the other versions once Fury is released without the legacy code.

**Comment by pksunkara** [comment]
Coveralls still seem to be wonky.

**Comment by pksunkara** [comment]
Merged. Please tag and release.
